File: bindings/ruby/lib/gelfgren.rb
# frozen_string_literal: true

require_relative "gelfgren/version"

begin
  # Try loading the native extension
  ruby_version = /(\d+\.\d+)/.match(RUBY_VERSION)
  require_relative "gelfgren/#{ruby_version}/gelfgren"
rescue LoadError
  require_relative "gelfgren/gelfgren"
end

module Gelfgren
  class Error < StandardError; end

  # Module methods for convenience
  class << self
    # Get library version
    def version
      VERSION
    end

    # Create a Bernstein polynomial from coefficients
    def bernstein(coeffs, a: 0.0, b: 1.0)
      BernsteinPolynomial.new(coeffs, a, b)
    end

    # Create a rational function from numerator and denominator polynomials
    def rational(numerator, denominator)
      RationalFunction.new(numerator, denominator)
    end

    # Create a Padé approximant from power series
    def pade_from_series(coeffs, n, m, a: -1.0, b: 1.0)
      PadeApproximant.from_power_series(coeffs, n, m, a, b)
    end

    # Create a Padé approximant from derivative values
    def pade_from_derivatives(derivs, n, m, x0: 0.0, a: -1.0, b: 1.0)
      PadeApproximant.from_derivatives(derivs, n, m, x0, a, b)
    end
  end
end

File: bindings/r/R/gelfgren.R
#' Gelfgren: Piecewise Rational Interpolation
#'
#' @description
#' High-performance piecewise rational interpolation and approximation library.
#' Implements algorithms from Jan Gelfgren's 1975 paper using Bernstein polynomial
#' representation for numerical stability.
#'
#' @docType package
#' @name gelfgren-package
#' @useDynLib gelfgren, .registration = TRUE
#' @import methods
NULL

#' Create a Bernstein polynomial
#'
#' @param coeffs Numeric vector of unscaled Bernstein coefficients
#' @param a Left endpoint of interval
#' @param b Right endpoint of interval
#' @return BernsteinPolynomial object
#' @export
#' @examples
#' # Create polynomial P(x) = 1 + 2x + 3x²
#' p <- bernstein_polynomial(c(1, 2, 3), 0, 1)
#' print(p)
#'
#' # Evaluate at points
#' x <- seq(0, 1, by = 0.25)
#' y <- p$evaluate(x)
#' plot(x, y, type = "b", main = "Bernstein Polynomial")
bernstein_polynomial <- function(coeffs, a, b) {
    BernsteinPolynomial$new(coeffs, a, b)
}

#' Create a rational function
#'
#' @param numerator BernsteinPolynomial for P(x)
#' @param denominator BernsteinPolynomial for Q(x)
#' @return RationalFunction object
#' @export
#' @examples
#' num <- bernstein_polynomial(c(1, 1), 0, 1)
#' den <- bernstein_polynomial(c(1, 2), 0, 1)
#' r <- rational_function(num, den)
#'
#' x <- seq(0, 1, by = 0.1)
#' y <- r$evaluate(x)
rational_function <- function(numerator, denominator) {
    RationalFunction$new(numerator, denominator)
}

#' Create a Padé approximant
#'
#' @param coeffs Numeric vector of power series coefficients
#' @param n Numerator degree
#' @param m Denominator degree
#' @param center Expansion center
#' @param a Left endpoint of interval
#' @param b Right endpoint of interval
#' @return PadeApproximant object
#' @export
#' @examples
#' # Approximate exp(x) with [2/2] Padé
#' coeffs <- c(1, 1, 0.5, 1/6, 1/24)
#' pade <- pade_approximant(coeffs, 2, 2, 0, -1, 1)
#'
#' x <- seq(-1, 1, by = 0.25)
#' approx <- pade$evaluate(x)
#' exact <- exp(x)
#' plot(x, exact, type = "l", col = "blue", main = "Padé vs exp")
#' points(x, approx, col = "red", pch = 19)
pade_approximant <- function(coeffs, n, m, center, a, b) {
    PadeApproximant$new(coeffs, as.integer(n), as.integer(m), center, a, b)
}

#' Create a uniform mesh
#'
#' @param a Left endpoint
#' @param b Right endpoint
#' @param n Number of subintervals
#' @return Mesh object
#' @export
mesh_uniform <- function(a, b, n) {
    Mesh$uniform(a, b, as.integer(n))
}

#' Create a Chebyshev mesh
#'
#' @param a Left endpoint
#' @param b Right endpoint
#' @param n Number of subintervals
#' @return Mesh object
#' @export
mesh_chebyshev <- function(a, b, n) {
    Mesh$chebyshev(a, b, as.integer(n))
}
